### ISSUE TITLE
Datepicker return type/0.19 support

### ISSUE THREAD

**Issue by jorgengranseth** [dwyl/elm-datepicker#32]
In its current iteration the Datepicker returns an Elm 0.18 `Date` which adds the user's local time zone. The user's time zone is an unnecessary complication, and I believe a neutral `(year, month, day)` of some sort would be easier to work with for most applications.

Since 0.19 deprecates the core `Date` API in favour of `elm/time`'s `Posix`, this is a good opportunity to consider whether the Datepicker API should return a different type.

What are your thoughts on the subject?

**Comment by nelsonic** [comment]
@jorgengranseth the only reason to take the User's time zone into account is when we want to default to "today" when presenting the datepicker UI (_and restrict date selection to dates in the future..._).
I _agree_ that we need to update to `0.19` ASAP so that we can migrate the projects which are using this package.

**Comment by jorgengranseth** [comment]
Does the `Datepicker` not allow you to pick a date in the past in its current iteration? Sounds like a sensible default, but it should also be possible to disable(?). Regardless you have a point that that might be a use case for time zones in the `Datepicker`.

I'll make an attempt this weekend to upgrade it to 0.19 with `Posix` first, and then assess the 0.19 API.

**Comment by nelsonic** [comment]
@jorgengranseth yes, the datepicker _does_ allow selecting a date in the past in it's _current_ form.
but one of the ideas was to allow us to _restrict_ this.

**Comment by nelsonic** [comment]
@jorgengranseth you have write access: https://github.com/dwyl/elm-datepicker/invitations
please create a PR for v0.19 if you can. 👍

**Comment by jorgengranseth** [comment]
I got all the examples working with a minimal diff port to 0.19 with `elm/time` on the `elm-0.19` branch. Let's see if I can port the tests too.. 😅 Feel free to give feedback on changes you would like in the meantime.

I think it would be a good idea to attempt porting from `elm/time` to a simpler type as suggested initially. There's also some refactoring I think would be sensible to make, and possibly drop `weekdayFormat` from the `Config` since we may as well provide the default formatter and let users specify their own function if they want anything else than `ddd`. But those are perhaps for other PRs to keep things clean?

**Comment by SimonLab** [comment]
Thanks for the PR #33 :+1: 
I think we'll need to do some tidy up on the releases. Not all the package versions have been published on https://package.elm-lang.org/ as the latest one there is 2.2.0
![image](https://user-images.githubusercontent.com/6057298/74637761-e941ac80-516a-11ea-8f76-03fe68c39311.png)

Running `elm bump` returns the following error:
![image](https://user-images.githubusercontent.com/6057298/74637802-fb234f80-516a-11ea-995b-54faa80e532a.png)

So I'm thinking first to push the two previous versions, 3.0.0 then 3.0.1 and hopefully this should allow us to create a new version with `elm bump`.
I don't know if there is a more direct way to do this, a `force publish` maybe?

**Comment by holgerl** [comment]
Any progress on the release? We are waiting for the new release with elm 0.19 support

**Comment by jorgengranseth** [comment]
To address this issue: Decide whether the API keeps `Date` from `justinmimbs/date` or translates to and from e.g.
```
type alias Date =
    { year : Int
    , month: Month
    , day : Int
    }
```
internally (using `Month` from `elm/time` and [`Date.fromCalendarDate`](https://package.elm-lang.org/packages/justinmimbs/date/latest/Date#fromCalendarDate)).

And consider other breaking changes to the API;
- Can the second argument (`monthData`) be a config option rather than direct input? Given an arbitrary `initMonth : { year : Int, month : Month }`, what prevents us from deducing what to show with the `validDate` function? Currently the two exposed functions to obtain `MonthData` are for the current and next month, and they are specifically for use in the `TwoMonth` example(?) If there is no actual flexibility in what to display this clutters the API.
- If not; should `showCalendar` take a record rather than separate arguments for readability?
  ```
    DatePicker.showCalendar
        { datePicker = model.calendar
        , monthData = DatePicker.getMonth model.calendar
        , config = config
        }
        |> Html.map DatePickerMsg
  ```
  If `monthData` is removed as an input argument, typical use could look like this, which is also pretty clean:
  ```
  datePickerConfig
      |> DatePicker.showCalendar model.calendar
      |> Html.map DatePickerMsg
  ```

**Comment by jorgengranseth** [comment]
@nelsonic mind if I release this to https://package.elm-lang.org/? Otherwise I'll make a fork.

**Comment by nelsonic** [comment]
Best person to ask is @SimonLab; it should definitely be shipped so others can use it. 👍

**Comment by SimonLab** [comment]
I've run the examples on my localhost as it has been a while I haven't use this package and haven't noticed any issues.
@jorgengranseth I think it's a good idea to publish it :+1:

**Comment by jorgengranseth** [comment]
To conclude on my questions from 2020-02-19 I made a PR #41 to clean up the use of MonthData. If that seems like a sensible change I can try publishing the package after merge since there are breaking API changes there.

**Comment by SimonLab** [comment]
I'm looking at resetting the version to 3.0.0. I'll use `elm bump` from the latest published version (2.2.0) to make sure that we have the correction versioning, then push the elm.json changes and create the tag. We might need to delete the two latest tags on Github (3.0.0  and 3.0.1), these two versions are not used/published so it shouldn't create any conflicts
